_posts/2019-03-19-til-validations-and-constraints.md

```diff
@@ -0,0 +1,229 @@
+I didn't. In the process of building a side project, I ran into the problem a few
+times! Let's talk about the goal of them and see the differences. We'll dive into
+**why we need both each** towards the end.
+
+## Data Integrity is Rule #1
+> Data Integrity is the maintenance of, and the assurance of the accuracy and
+> consistency of, data over its entire lifecycle
+> - [Wikipedia](https://en.wikipedia.org/wiki/Data_integrity)
+
+So we're building a super cool app with users that can login and logout.
+We'd probably have some schema like this:
+```elixir
+# Using Phoenix 1.4 with Contexts but still applies all the same
+defmodule MyCoolWebApp.Accounts.User do
+  use Ecto.Schema
+  import Ecto.Changeset
+
+  schema "users" do
+    field :display_name, :string
+    field :email, :string
+    field :password_hash, :string
+    field :password, :string, virtual: true
+
+    timestamps()
+  end
+end
+```
+
+And inside of it, we'd want to describe the `changeset`:
+```elixir
+def changeset(user, attrs) do
+  user
+  |> cast(attrs, [:display_name, :email, :password])
+end
+```
+
+If we only had this, we might have a few headaches come our way. It is very easy to
+prematurely submit a form without filling all the fields out. Potentially, now the
+user's profile has a lovely email or password of `nil`, or worse `""`, in the
+database. This would suck, for the user(s) and the developer(s).
+
+So in order to fix, we'd use a validation!
+
+### Validations
+Many of the validations we have in Ecto will be executed without needing
+to interact with the database. That means the validation will be executed prior to the
+attempt of inserting or updating something in the database. If we wanted to
+insert a new user into our database, we'd first want to make sure there is data
+inside of the changeset.
+
+Let's add the [`validate_require/3`](https://hexdocs.pm/ecto/Ecto.Changeset.html#validate_required/3) into our `changeset`:
+```elixir
+def changeset(user, attrs) do
+  user
+  |> cast(attrs, [:display_name, :email, :password])
+  |> validate_required([:display_name, :email, :password])
+end
+```
+
+And for free, we get a set of errors coming out if a user were to make a mistake:
+```elixir
+iex(1)> %User{} |> User.changeset(%{})
+#Ecto.Changeset<
+  action: nil,
+  changes: %{},
+  errors: [
+    display_name: {"can't be blank", [validation: :required]},
+    email: {"can't be blank", [validation: :required]},
+    password: {"can't be blank", [validation: :required]}
+  ],
+  data: #MyCoolWebApp.Accounts.User<>,
+  valid?: false
+>
+```
+
+There are a ton of validations out there that can enhance your app! Take a
+look at the documentation for the [Ecto.Changeset](https://hexdocs.pm/ecto/Ecto.Changeset.html#summary). Now the next
+phase will be looking at when we need to apply some constraints.
+
+### Constraints
+You might have gotten to this point and thought, why are there differences?
```

Comment:
I would phrase this a little differently. "If we use validations, why do we also need constraints?"